darshan-runtime/lib/darshan-ldms.c

```diff
@@ -0,0 +1,270 @@
+    if (getenv("POSIX_ENABLE_LDMS"))
+        dC.posix_enable_ldms = 0;
+    else
+        dC.posix_enable_ldms = 1;
+
+    if (getenv("MPIIO_ENABLE_LDMS"))
+        dC.mpiio_enable_ldms = 0;
+    else
+        dC.mpiio_enable_ldms = 1;
+
+    /* Disable STDIO if verbose is enabled to avoid a recursive
+    function for darshan_ldms_connector_send() */
+    if (getenv("STDIO_ENABLE_LDMS"))
+        if (!getenv("DARSHAN_LDMS_VERBOSE"))
+            dC.stdio_enable_ldms = 0;
+        else
+            dC.stdio_enable_ldms = 1;
+    else
+        dC.stdio_enable_ldms = 1;
+    
+    if (getenv("HDF5_ENABLE_LDMS"))
+        dC.hdf5_enable_ldms = 0;
+    else
+        dC.hdf5_enable_ldms = 1;
+
+    if (!getenv("DARSHAN_LDMS_STREAM"))
+        dC.env_ldms_stream = "darshanConnector";
+    
+    const char* env_ldms_xprt    = getenv("DARSHAN_LDMS_XPRT");
+    const char* env_ldms_host    = getenv("DARSHAN_LDMS_HOST");
+    const char* env_ldms_port    = getenv("DARSHAN_LDMS_PORT");
+    const char* env_ldms_auth    = getenv("DARSHAN_LDMS_AUTH");
+
+    /* Check/set LDMS transport type */
+    if (!env_ldms_xprt || !env_ldms_host || !env_ldms_port || !env_ldms_auth){
+        printf("Either the transport, host, port or authentication is not given\n");
+        return;
+    }
+
+    pthread_mutex_lock(&dC.ln_lock);
+    dC.ldms_darsh = setup_connection(env_ldms_xprt, env_ldms_host, env_ldms_port, env_ldms_auth);
+    if (dC.conn_status != 0) {
+        printf("Error setting up connection to LDMS streams daemon: %i -- exiting\n", dC.conn_status);
+        pthread_mutex_unlock(&dC.ln_lock);
+        return;
+    }
+    else if (dC.ldms_darsh->disconnected){
+        printf("Disconnected from LDMS streams daemon -- exiting\n");
+        pthread_mutex_unlock(&dC.ln_lock);
+        return;
+    }
+    pthread_mutex_unlock(&dC.ln_lock);
+    return;
+}
+
+void darshan_ldms_connector_send(uint64_t record_id, int64_t rank, int64_t record_count, char *rwo, int64_t offset, int64_t length, int64_t max_byte, int64_t rw_switch, int64_t flushes,  double start_time, double end_time, double total_time, char *mod_name, char *data_type)
+{
+    char jb11[1024];
+    int rc, ret, i, size, exists;
+    const char *filepath;
+    struct timespec tspec_start, tspec_end;
+    uint64_t micro_s;
+    
+    dC.env_ldms_stream  = getenv("DARSHAN_LDMS_STREAM");
+
+    pthread_mutex_lock(&dC.ln_lock);
+    if (dC.ldms_darsh != NULL)
+        exists = 1;
+    else
+        exists = 0;
+    pthread_mutex_unlock(&dC.ln_lock);
+
+    if (!exists){
+        return;
+    }
+
+    /* get the full file path from record ID */
+    filepath = darshan_core_lookup_record_name(record_id);
+
+    if (strcmp(rwo, "open") == 0)
+        dC.open_count = record_count;
+
+    /* set record count of closes to number of opens since we are closing the same file we opened.*/
+    if (strcmp(rwo, "close") == 0)
+        record_count = dC.open_count;
```

Comment:
Yes I see what you mean. The goal was to keep track of the number of closes that occurred (i.e. counter) for each file. For example, if there are "x" number of opens of a specific file then, by the end of the run, there should be "x" number of closes of the same file.

I tried searching through darshan's module code but couldn't find any counters for this I/O event. Does darshan have a counter that keeps track of the number of file closes?  If not, then I was going to try using the record_id and see if it's possible to map the cached ``open_count`` value to the corresponding file.

Also, I'm assuming that all applications close any/all files that were opened before completing (which could be completely wrong). If I am wrong, then I'll just delete this block of code and set it to "-1" (i.e. N/A).

Reply:
Yep, you're right, we don't have any counters keeping track of closes, which complicates a bit what you're trying to do. The reasoning for that on our end was mostly just that it's not a super interesting counter -- it will almost always match the open counter. We do keep track of how much time is spent in close operations (and attribute this to `META_TIME` timer), but that's basically it.

If you really want to capture it, probably you'll have to do something along the lines of what you mention above where you maintain your own map from file record ID to the close count in LDMS code. I was mostly just concerned about the possibility of these erroneous results confusing folks (e.g., this file was opened 1 time but it says it was closed 1000 times...what gives? oh, it's accidentally the counter from another file), so as long as that's cleared up I'm fine with whichever way you want to go here (punt on this functionality or implement it in a separate mapping).

FWIW, I would expect the vast majority of apps to be pretty good about closing each file they open. That said, not closing a file won't likely have negative side effects on most applications, so folks that don't do it (either because they forgot or because they are lazy) aren't really punished anyway. Once the application terminates, the OS will implicitly close all opened files for you. There are OS limits to the number of opened files you can have, so not closing can affect you if you are up against that, but in general forgetting to call it isn't going to result in data loss or anything, so not always that critical.

Reply:
This helps a lot (better understanding of how the OS works) so thank you for the information!

Our LDMS team already have some python analyses that query/use the close counts so I figured it'd be easier to continue collecting this information instead of updating the analyses. I mapped the record_id to the file name via 2 dynamic arrays and it seems to work as expected (ran a test with IOR and mpi-io-test in ``darshan-test/``).

There was a good chunk of code that I added so if you can please review this and see if there're any issues then that would be great!

Reply:
I just looked this over. So, to summarize the current code, it seems that you are caching the number of file opens for each file record on `open` operations, and then using that value for the record count on `close` operations.

I still think that's not quite right. Correct me if I'm wrong, but I'm guessing what you really want to provide to LDMS here is the total number of observed close operations so far for the given file? Since Darshan doesn't track that, I think we would need to track that independently and not use the open count as a substitute for this value as the current code does. Assuming that all sounds right, I think the easiest way to do this is to add a value to the `*_record_ref` structure for close counts, increment that value on close operations, and then pass that value to LDMS send routine? 

E.g., for POSIX:

```
struct posix_file_record_ref
{
    ...
#ifdef HAVE_LDMS
    int64_t close_counts;
#endif
}; 
```

Then in the LDMS code block in the close wrapper, do something like:

```
#ifdef HAVE_LDMS
        rec_ref->close_counts++;
        /* publish close information for posix */
        extern struct darshanConnector dC;
        if(dC.ldms_lib)
            if(dC.posix_enable_ldms)
                darshan_ldms_connector_send(rec_ref->file_rec->base_rec.id, rec_ref->file_rec->base_rec.rank, rec_ref->close_counts, "close", -1, -1, -1, -1, -1, tm1, tm2, rec_ref->file_rec->fcounters[POSIX_F_META_TIME], "POSIX", "MOD");
#endif
```

Does that make sense or am I missing something?

Reply:
Yes that is exactly what I am trying to do and this will work! It was stuck in my mind that the only solution was use ``record_id`` to map the number of opens & closes (which made it much more complicated and confusing than it needed to be) so I didn't even think about this much more simpler/easier way.

I've added these changes and, after testing, the correct number of closes are being collected. Thank you for double checking this!

Reply:
Thanks for having a look, I think this looks good now.